### PR TITLE
[PowerShell] Add PowerShell API client generator (WIP)

### PR DESCRIPTION
**Pull request by wing328**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [x] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

Closes https://github.com/swagger-api/swagger-codegen/issues/4320

It's based on the PowerShell petstore sample provided by @beatcracker

It's still work-in-progress.

### CODE REVIEW

**Comment by wing328** [comment]
@beatcracker I've a couple of questions if you don't mind.

1. Do we need to have 1 function per 1 PS file? Currently we generate functions for different endpoints within the same PS file.

2. Is it ok to put the API-related files under the "API" subfolder and model files under the "Model" subfolder? I can put all of them under "Public" but just want to see if we can use subfolders to organize these files.

**Comment by beatcracker** [comment]
@wing328 

> Do we need to have 1 function per 1 PS file? Currently we generate functions for different endpoints within the same PS file.

No, absolutely not. It's OK to bundle them together.

> Is it ok to put the API-related files under the "API" subfolder and model files under the "Model" subfolder? I can put all of them under "Public" but just want to see if we can use subfolders to organize these files.

The same as above. Private/Public folder structure is just an unofficial convention which helps keep things logically separated when writing code manually. We can organize code the way we see fit.

---

I'm currently testing your PR locally and I'll give you more feedback soon.

**Comment by beatcracker** [comment]
Ok, here is today's results:

1. I've fixed some essential issues with mustache templates. I'm not sure what's the best way to merge them with this PR, so for now I've just created [beatcracker/powershell_generator_fix](https://github.com/beatcracker/swagger-codegen/tree/powershell_generator_fix) branch in my fork.

    There are still more issues, which currently I don't know how to fix. I'll try to point to them in this PR's code review later.

2. Regarding one-function-per-file question. I've said, that there is no hard requirement for functions to be in separate files, but I've forgot that there is a code in the manifest generator that generates `FunctionsToExport` property using these filenames. So we need either to fallback to separate files or figure out some other way.

**Comment by wing328** [comment]
1. I've merged your fix into my branch via 4850b83

> So we need either to fallback to separate files or figure out some other way.

I think we need to figure some other ways as it's not easy to generate an endpoint function per file.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        )
+    }
+}
+
+function PetApi-deletePet {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = , ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
+        [IO.Swagger.Model.Pet]
+        ${ petId }
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Parameter definitions should be separated by comma, e.g.: `${petId}, ${apiKey}`

Extended example (note the comma after ``${petId}`):

```posh
        [Parameter(ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
        [IO.Swagger.Model.Pet]
        ${petId},

        [Parameter(ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $false)]
        [IO.Swagger.Model.Pet]
        ${apiKey}
```

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+function PetApi-updatePetWithForm {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = , ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
+        [IO.Swagger.Model.Pet]
+        ${ petId }
+        [Parameter(Position = , ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $false)]
+        [IO.Swagger.Model.Pet]
+        ${ name }
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: `$name` should have type of `string`. In PoweShell, you can use either [accelerator](https://learn-powershell.net/2013/10/10/quick-hits-finding-creating-and-removing-powershell-type-accelerators/) type, such as `[string]` or `[xml]` or .NET type, such as `[System.String]` or `[System.Int64]`.

File: samples/client/petstore/powershell_test/src/IO.Swagger/IO.Swagger.psm1 [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+#endregion
+
+
+#region Initialize APIs
+
+'Creating object: IO.Swagger.Api.PetApi' | Write-Verbose
+$Script:PetApi= New-Object -TypeName IO.Swagger.Api.PetApi -ArgumentList @($null)
+
+'Creating object: IO.Swagger.Api.PetApi' | Write-Verbose
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/IO.Swagger.psm1]: Look like something gone wrong with mustache template, so it repeats the same line.

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Order.ps1 [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        [System.Nullable[Int64]]
+        ${ id },
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[Int64]]
+        ${ petId },
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[Int32]]
+        ${ quantity },
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[JavaTimeLocalDateTime]]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Order.ps1]: Wrong type mapping slipped in. This should probably be `[System.Nullable[System.DateTime]]`.

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Pet.ps1 [hunk 5]
```diff
@@ -0,0 +1,36 @@
+function New-Pet {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[Int64]]
+        ${ id },
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[Category]]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Pet.ps1]: This, and others like this, should use full type name, e.g.: `[System.Nullable[IO.Swagger.Model.Category]]`

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-ApiResponse.ps1 [hunk 6]
```diff
@@ -0,0 +1,24 @@
+function New-ApiResponse {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[Int32]]
+        ${ code },
+        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true)]
+        [System.Nullable[String]]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-ApiResponse.ps1]: `String` can't be nullable, got this exception in the PowerShell: 

```none
Unable to find type [System.Nullable[System.String]].
Details: GenericArguments[0], 'System.String', on 'System.Nullable`1[T]' violates the constraint of type 'T'.
```

**Comment by beatcracker** [comment]
> I think we need to figure some other ways as it's not easy to generate an endpoint function per file.

Can we obtain this info from the generator? `operationId` and `classname` tags in mustache look promising.

If not, I can always write some code to parse `*.ps1` files for function names.

File: modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+$Manifest = @{
+    Path = "$ScriptDir\src\{{{packageName}}}\{{{packageName}}}.psd1"
+
+    Author = 'Swagger Codegen Team'
+    CompanyName = 'swagger.io'
+    Description = '{{{packageName}}} - the PowerShell module for {{{appName}}}'
+
+    RootModule = '{{{packageName}}}.psm1'
+    Guid = 'a27b908d-2a20-467f-bc32-af6f3a654ac5' # Has to be static, otherwise each new build will be considered different module
```
Comment (beatcracker) [inline modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache]: While this has to be static between module builds, it has to be unique for a given API. I.e, we need deterministic GUIDs. May we could use something like this in the build script: [How can I generate a GUID for a string?](https://stackoverflow.com/a/28608877/4424236).

**Comment by beatcracker** [comment]
We should also consider naming scheme for the functions. For example, `PetApi-addPet` is not a vaild PowerShell function name, which should follow `Verb-Noun` pattern and use [approved verbs](https://msdn.microsoft.com/en-us/library/ms714428.aspx). Invalid function names for exported functions will generate warnings on module import.

Perhaps, we could use something like `Invoke-AddPet`.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 [hunk 8]
```diff
@@ -0,0 +1,164 @@
+function PetApi-addPet {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = , ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Template failed to generate `Position` attribute correctly (missing position number after `=`)

**Comment by wing328** [comment]
@beatcracker thanks for all the feedback 👍  Working on it...

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 1 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Fixed via 3902199

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 2 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: I used the primitive type defined in https://github.com/swagger-api/swagger-codegen/pull/5789/files#diff-f766d24f8851685d039700dd876127ceR44. Is that ok?

File: samples/client/petstore/powershell_test/src/IO.Swagger/IO.Swagger.psm1 (hunk 3 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/IO.Swagger.psm1]: Fixed via 3902199

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Order.ps1 (hunk 4 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Order.ps1]: Fixed via 3902199

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Pet.ps1 (hunk 5 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-Pet.ps1]: Fixed via 3902199

File: samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-ApiResponse.ps1 (hunk 6 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/Model/New-ApiResponse.ps1]: Fixed via fae84f8

File: modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache (hunk 7 above)
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache]: For this, we can reuse the same logic from C# server stub generator, which also requires a unique GUID for project/solution file. I'll do it in my next update.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 8 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: I tried to fix it and now the position starts at 1. Is that ok or it must be started at 0?

**Comment by wing328** [comment]
> Perhaps, we could use something like Invoke-AddPet.

What about "Invoke-PetApi-addPet" or "Invoke-PetApiAddPet" as I've seen users putting the same operationId (e.g. getName) in more than 1 endpoints?

(FYI. "operationId" must be unique according to OpenAPI spec 2.0 but there are specs out there not conforming to this constraint)

**Comment by wing328** [comment]
> Can we obtain this info from the generator? operationId and classname tags in mustache look promising.

Sorry what info do you need in order to achieve more than 1 function per PS file?

**Comment by wing328** [comment]
In this [PowerShell file](https://github.com/MeshkDevs/InvokeTwitterAPIs/blob/master/InvokeTwitterAPIs.psm1), I found more than 1 function:

- https://github.com/MeshkDevs/InvokeTwitterAPIs/blob/master/InvokeTwitterAPIs.psm1#L123
- https://github.com/MeshkDevs/InvokeTwitterAPIs/blob/master/InvokeTwitterAPIs.psm1#L196
and more.

Not sure if it's the right way to do it or there are any hidden issues with this approach.

**Comment by beatcracker** [comment]
> `Invoke-PetApiAddPet`

I was thinking along those lines and while this doesn't please my eyes, it looks as close to convention as we can afford for now. I think we should stick with it, at least for foreseeable future.

> Sorry what info do you need in order to achieve more than 1 function per PS file?

I hadn't looked closely at the generator code, but as far as I understand, it reads definitions from the swagger spec and feeds them to the mustache which produces actual code using templates. I had an idea that we could include mustache template in `Build.ps1.mustache` that will render the list of the functions (like `api.mustache`) . Something like this (probably not valid, but just to show an idea):

* `Build.ps1.mustache`
```posh
$Manifest = @{
    FunctionsToExport  = @(
        {{#operations}}
        {{#operation}}
            '{{{classname}}}-{{{operationId}}}'
        {{/operation}}
        {{/operations}}
    )
}
```

> Not sure if it's the right way to do it or there are any hidden issues with this approach.

There is nothing wrong with this approach, this is a perfectly valid PowerShell module, albeit a slightly ungainly. 

Let me try to explain things from the historical perspective:

---

Originally PowerShell hadn't a package repository. This meant that people would package/install modules and functions as they saw fit. Later, the lack of the central repository, became severe limiting factor, since languages like perl/python/.NET make package installation easy and PowerShell fell behind. This and the advancement of the CI/CD pipelines eventually lead to the creation of the [PowerShell Gallery](https://www.powershellgallery.com/), which is based on the NuGet and provides discovery, installation and versioning for modules/scripts.

So without the PowerShell Gallery, you could just put all your functions in the `psm1` file, optionally supply `psd1` manifest and call it a day. One would install module by copying module folder to the PowerShell system module directory or just importing it manually using `Import-Module` cmdlet. Often, modules included install scripts to automate that kind of tasks (like [Install.ps1](https://github.com/MeshkDevs/InvokeTwitterAPIs/blob/master/Install.ps1)).

But now, we have PowerShell Gallery, which strives to deliver high-quality code and has reasonable [publishing guidelines](https://msdn.microsoft.com/en-us/powershell/gallery/psgallery/psgallery-publishingguidelines).

I feel, that as a long-term goal, swagger-codegen should be able to generate PowerShell modules that are fit to to be published there. Ideally, I'd like to be able to setup CI/CD pipeline that would automatically generate a module from swagger spec, run tests and publish it to the PS Gallery

And here we come to the " one-function-per-file" question. As I've said, module manifest should expose `FunctionsToExport` to make those functions discoverable by search in the PS Gallery. The easiest way to do this is to have one function per file, with the file named after the function. This way one could just enumerate directory contents during module build and update module manifest accordingly.

But it's not the "one-and-only" right way. Since we generate the module and have our own constraints on how we can do it, we can produce any acceptable file/folder structure. We just have to make sure that it fits the definition of the [well-formed module](https://msdn.microsoft.com/en-us/library/dd878350.aspx) and adheres to the PS Gallery [publishing guidelines](https://msdn.microsoft.com/en-us/powershell/gallery/psgallery/psgallery-publishingguidelines).

---

#### TL;DR

It's nice to have a good module which adheres to (un)offical standards and having `FunctionsToExport` is a strong requirement. It doesn't mater how exactly we generate it, it just have to be there.

**Comment by wing328** [comment]
> I was thinking along those lines and while this doesn't please my eyes, it looks as close to the convention as we can afford for now. I think we should stick with it, at least for foreseeable future.

Done via 0f0a586

**Comment by wing328** [comment]
> I feel, that as a long-term goal, swagger-codegen should be able to generate PowerShell modules that are fit to to be published there. Ideally, I'd like to be able to setup CI/CD pipeline that would automatically generate a module from swagger spec, run tests and publish it to the PS Gallery

Yup, I think that's the ultimate goal.

> We just have to make sure that it fits the definition of the well-formed module and adheres to the PS Gallery publishing guidelines.

Totally agreed the auto-generated PowerShell module needs to conform to the guideline.

**Comment by beatcracker** [comment]
Here is the new branch with more fixes: [powershell_generator_fix_1 ](https://github.com/beatcracker/swagger-codegen/tree/powershell_generator_fix_1)

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = 1, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
+        [Int64]
+        ${petId},
+        [Parameter(Position = 2, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $false)]
+        [String]
+        ${additionalMetadata},
+        [Parameter(Position = 3, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $false)]
+        [IO.Swagger.Model.String]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Should be just plain `[String]`.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 [hunk 10]
```diff
@@ -0,0 +1,164 @@
+function Invoke-PetApiAddPet {
+    [CmdletBinding()]
+    Param (
+        [Parameter(Position = 1, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
```
Comment (beatcracker) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Position should start from `0`. Unfortunately, I couldn't figure out how to do this with `jmustache`.

Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: No worry. I'll take care of it

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 9 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: You're right. Will fix it.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 10 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: This's fixed. Will push it out later this weekend.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 9 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Couldn't find a fix at the moment. Will give it another try over the weekend.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 10 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Pushed the fix.

File: samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1 (hunk 9 above)
Comment (wing328) [inline samples/client/petstore/powershell_test/src/IO.Swagger/API/PetApi.ps1]: Found a fix and pushed the commit https://github.com/swagger-api/swagger-codegen/pull/5789/commits/dc31206a7afb0146187163f92f5ea66f21150e17

**Comment by wing328** [comment]
Merged https://github.com/beatcracker/swagger-codegen/tree/powershell_generator_fix_1 into my branch.

File: modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache (hunk 7 above)
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache]: Didn't have a chance to work on this yet.

Will do it later today or tomorrow.

**Comment by beatcracker** [comment]
Great, I'll try to check updated generator towards the evening.

File: modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache (hunk 7 above)
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/powershell/Build.ps1.mustache]: @beatcracker UPDATE: added GUID support via 50b8795

**Comment by beatcracker** [comment]
> Great, I'll try to check updated generator towards the evening.

Well, that didn't work out, too busy 😞. I'll try to have a look at the new code during the weekend.

**Comment by wing328** [comment]
> Well, that didn't work out, too busy 😞. I'll try to have a look at the new code during the weekend.

No worry. Please review the change when you're free.

**Comment by beatcracker** [comment]
Managed to build and test latest version. There is still one issue in the `New-Pet.ps1`: double `[` in the parameter type. I couldn't find relevant code in the generator. After fixing this issue manually, I'm able to import module and make API calls using generated cmdlets.

```posh
[Parameter(Position = 0, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
[[String]]
${photoUrls},
```

I haven't tested all commands, since this is time consuming. I could provide template for basic [Pester](https://github.com/pester/Pester) tests so they could be generated alongside with the code.

Unfortunately, it looks like Petstore API allows anyone to change stored data, so it would be hard to test against (e,g, `Invoke-PetApiGetPetById` requires fixed `Id` and there is no way PetStore API can provide it, since anyone can add/delete pets).

Usually this is solved with mocking `Invoke-WebRrequest`  in Pester, but since this module is using binaries this approach is not available.

**Comment by wing328** [comment]
> double [ in the parameter type. I couldn't find relevant code in the generator

so `[[String]]` needs to be replaced as `[String]`, right? I'll take care of that.

>  I could provide template for basic Pester tests so they could be generated alongside with the code.

Good idea. We can generate test files for each Api class (e.g. PetApi.ps1) for easier testing. I will generate some very basic files to start with.

> Unfortunately, it looks like Petstore API allows anyone to change stored data, so it would be hard to test against

In our CIs (e.g. AppVeyor for testing C# API clients), we run a local Petstore server so bad data in the public server won't affect our test.

**Comment by beatcracker** [comment]
> [[String]] needs to be replaced as [String], right?

Yes.

> Good idea. We can generate test files for each Api class (e.g. PetApi.ps1) for easier testing. 

Here is the basic test example for Pester:

```posh
Describe 'IO.Swagger PetStore API' {
    Context 'PetApi' {
        It 'Invoke-PetApiGetPetById' {
            $ret = Invoke-PetApiGetPetById -petId 0
            $ret | Should BeOfType IO.Swagger.Model.Pet
            $ret.id | Should Be 0
        }
    }

    Context 'StoreApi' {
        It 'Invoke-StoreApiGetOrderById' {
            $ret = Invoke-StoreApiGetOrderById -orderId 0
            $ret | Should BeOfType IO.Swagger.Model.Order
            $ret.Id | Should Be  0
        }
    }
}
```

**Comment by wing328** [comment]
Using the following as an example
``` 
        [Parameter(Position = 0, ValueFromPipelineByPropertyName = $true, Mandatory = $true)]
        [[String]]
        ${photoUrls},
```
The parameter is an array of string. If we simply put [String], how can Powershell distinguish between array of string and just string?

**Comment by beatcracker** [comment]
> The parameter is an array of string. If we simply put [String], how can Powershell distinguish between array of string and just string?

For array of strings, it should look like this: `[String[]]`

**Comment by wing328** [comment]
> For array of strings, it should look like this: [String[]]

Thanks. I'll fix it.

**Comment by beatcracker** [comment]
Almost there, now we need to remove one set of square brackets at we'll get correct array notation: `[[String[]]]` --> `[String[]]`.

**Comment by beatcracker** [comment]
Yippee ki-yay, it builds and runs!

**Comment by wing328** [comment]
@beatcracker thanks for your tremendous effort 👏 👏 👏 . Finally we've a working PowerShell generator. There are some enhancements I can think of and I'll create a separate task to track those.

**Comment by beatcracker** [comment]
You're welcome, glad we finally did it.

> There are some enhancements I can think of and I'll create a separate task to track those.

I also plan to have a fresh look at the code/module structure to see if it can be improved.

**Comment by wing328** [comment]
Tweet to promote the new generator: https://twitter.com/wing328/status/877184938344239104

**Comment by wing328** [comment]
Also added @beatcracker as the template creator of PowerShell in the readme: https://github.com/swagger-api/swagger-codegen/blob/master/README.md#template-creator
